File: bugsnag-android-core/src/main/java/com/bugsnag/android/MetaData.java
package com.bugsnag.android;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Observable;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

/**
 * A container for additional diagnostic information you'd like to send with
 * every error report.
 * <p>
 * Diagnostic information is presented on your Bugsnag dashboard in tabs.
 */
public class MetaData extends Observable implements JsonStream.Streamable, MetaDataAware {

    @NonNull
    final Map<String, Object> store;
    final ObjectJsonStreamer jsonStreamer;

    MetaData() {
        this(new ConcurrentHashMap<String, Object>());
    }

    MetaData(@NonNull Map<String, Object> map) {
        store = new ConcurrentHashMap<>(map);
        jsonStreamer = new ObjectJsonStreamer();
    }

    @Override
    public void toStream(@NonNull JsonStream writer) throws IOException {
        jsonStreamer.objectToStream(store, writer);
    }

    @Override
    public void addMetadata(@NonNull String section, @Nullable String key, @Nullable Object value) {
        Map<String, Object> tab = getTab(section);
        setChanged();
        if (value != null) {
            tab.put(key, value);
            notifyObservers(new NativeInterface.Message(
                    NativeInterface.MessageType.ADD_METADATA,
                    Arrays.asList(section, key, value)));
        } else {
            tab.remove(key);
            notifyObservers(new NativeInterface.Message(
                    NativeInterface.MessageType.REMOVE_METADATA,
                    Arrays.asList(section, key)));
        }
    }

    @Override
    public void clearMetadata(@NotNull String section, @Nullable String key) {
        store.remove(section);
        setChanged();
        notifyObservers(new NativeInterface.Message(
                NativeInterface.MessageType.CLEAR_METADATA_TAB, section));
    }

    @Nullable
    @Override
    public Object getMetadata(@NotNull String section, @Nullable String key) {
        return null;
    }

    @NonNull
    private Map<String, Object> getTab(String tabName) {
        @SuppressWarnings("unchecked")
        Map<String, Object> tab = (Map<String, Object>) store.get(tabName);

        if (tab == null) {
            tab = new ConcurrentHashMap<>();
            store.put(tabName, tab);
        }

        return tab;
    }

    void setFilters(Collection<String> filters) {
        Collection<String> data = new HashSet<>(filters);
        jsonStreamer.filters.clear();
        jsonStreamer.filters.addAll(data);
    }

    Set<String> getFilters() {
        return jsonStreamer.filters;
    }

    @NonNull
    static MetaData merge(@NonNull MetaData... metaDataList) {
        List<Map<String, Object>> stores = new ArrayList<>();
        Collection<String> filters = new HashSet<>();
        for (MetaData metaData : metaDataList) {
            if (metaData != null) {
                stores.add(metaData.store);

                if (metaData.jsonStreamer.filters != null) {
                    filters.addAll(metaData.jsonStreamer.filters);
                }
            }
        }

        @SuppressWarnings({"unchecked", "rawtypes"})
        MetaData newMeta = new MetaData(mergeMaps(stores.toArray(new Map[0])));
        newMeta.setFilters(filters);
        return newMeta;
    }

    @SafeVarargs
    @NonNull
    private static Map<String, Object> mergeMaps(@NonNull Map<String, Object>... maps) {
        Map<String, Object> result = new ConcurrentHashMap<>();

        for (Map<String, Object> map : maps) {
            if (map == null) {
                continue;
            }

            // Get a set of all possible keys in base and overrides
            Set<String> allKeys = new HashSet<>(result.keySet());
            allKeys.addAll(map.keySet());

            for (String key : allKeys) {
                Object baseValue = result.get(key);
                Object overridesValue = map.get(key);

                if (overridesValue != null) {
                    if (baseValue instanceof Map && overridesValue instanceof Map) {
                        // Both original and overrides are Maps, go deeper
                        @SuppressWarnings("unchecked")
                        Map<String, Object> first = (Map<String, Object>) baseValue;
                        @SuppressWarnings("unchecked")
                        Map<String, Object> second = (Map<String, Object>) overridesValue;
                        result.put(key, mergeMaps(first, second));
                    } else {
                        result.put(key, overridesValue);
                    }
                } else {
                    if (baseValue != null) { // No collision, just use base value
                        result.put(key, baseValue);
                    }
                }
            }
        }

        return result;
    }
}

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Metadata.kt
@file:Suppress("UNCHECKED_CAST")

package com.bugsnag.android

import java.io.IOException
import java.util.HashMap
import java.util.HashSet
import java.util.concurrent.ConcurrentHashMap

/**
 * A container for additional diagnostic information you'd like to send with
 * every error report.
 *
 * Diagnostic information is presented on your Bugsnag dashboard in tabs.
 */
internal data class Metadata @JvmOverloads constructor(
    internal val store: MutableMap<String, Any> = ConcurrentHashMap(),
    val jsonStreamer: ObjectJsonStreamer = ObjectJsonStreamer(),
    val redactedKeys: Set<String> = jsonStreamer.redactedKeys
) : JsonStream.Streamable, MetadataAware {

    @Throws(IOException::class)
    override fun toStream(writer: JsonStream) {
        jsonStreamer.objectToStream(store, writer, true)
    }

    override fun addMetadata(section: String, value: Map<String, Any?>) {
        value.entries.forEach {
            addMetadata(section, it.key, it.value)
        }
    }

    override fun addMetadata(section: String, key: String, value: Any?) {
        if (value == null) {
            clearMetadata(section, key)
        } else {
            var tab = store[section]
            if (tab !is MutableMap<*, *>) {
                tab = ConcurrentHashMap<Any, Any>()
                store[section] = tab
            }
            insertValue(tab as MutableMap<String, Any>, key, value)
        }
    }

    private fun insertValue(map: MutableMap<String, Any>, key: String, value: Any) {
        var obj = value

        if (obj is MutableMap<*, *> && map.isNotEmpty()) {
            obj = mergeMaps(listOf(map as Map<String, Any>, value as Map<String, Any>))
        }
        map[key] = obj
    }

    override fun clearMetadata(section: String) {
        store.remove(section)
    }

    override fun clearMetadata(section: String, key: String) {
        val tab = store[section]

        if (tab is MutableMap<*, *>) {
            tab.remove(key)

            if (tab.isEmpty()) {
                store.remove(section)
            }
        }
    }

    override fun getMetadata(section: String): Map<String, Any>? {
        return store[section] as (Map<String, Any>?)
    }

    override fun getMetadata(section: String, key: String): Any? {
        return when (val tab = store[section]) {
            is Map<*, *> -> (tab as Map<String, Any>?)!![key]
            else -> tab
        }
    }

    fun toMap(): Map<String, Any> {
        val hashMap = HashMap(store)

        // deep copy each section
        store.entries.forEach {
            if (it.value is ConcurrentHashMap<*, *>) {
                hashMap[it.key] = ConcurrentHashMap(it.value as ConcurrentHashMap<*, *>)
            }
        }
        return hashMap
    }

    fun setRedactedKeys(redactKeys: Collection<String>) {
        val data = HashSet(redactKeys)
        jsonStreamer.redactedKeys.clear()
        jsonStreamer.redactedKeys.addAll(data)
    }

    companion object {
        fun merge(vararg data: Metadata): Metadata {
            val stores = data.map { it.toMap() }
            val redactKeys = data.flatMap { it.jsonStreamer.redactedKeys }
            val newMeta = Metadata(mergeMaps(stores))
            newMeta.setRedactedKeys(redactKeys.toSet())
            return newMeta
        }

        internal fun mergeMaps(data: List<Map<String, Any>>): MutableMap<String, Any> {
            val keys = data.flatMap { it.keys }.toSet()
            val result = ConcurrentHashMap<String, Any>()

            for (map in data) {
                for (key in keys) {
                    getMergeValue(result, key, map)
                }
            }
            return result
        }

        private fun getMergeValue(
            result: ConcurrentHashMap<String, Any>,
            key: String,
            map: Map<String, Any>
        ) {
            val baseValue = result[key]
            val overridesValue = map[key]

            if (overridesValue != null) {
                if (baseValue is Map<*, *> && overridesValue is Map<*, *>) {
                    // Both original and overrides are Maps, go deeper
                    val first = baseValue as Map<String, Any>?
                    val second = overridesValue as Map<String, Any>?
                    result[key] = mergeMaps(listOf(first!!, second!!))
                } else {
                    result[key] = overridesValue
                }
            } else {
                if (baseValue != null) { // No collision, just use base value
                    result[key] = baseValue
                }
            }
        }
    }

    fun copy() = this.copy(
        store = toMap().toMutableMap(),
        jsonStreamer = jsonStreamer,
        redactedKeys = redactedKeys
    )
}
